### ISSUE TITLE
Getting different probability distribution between R and PMML 

### ISSUE THREAD

**Issue by edenwind** [jpmml/jpmml-evaluator#128]
Hi there

Sorry for bothering you again. I have solved the missingValueException in the last issue [https://github.com/jpmml/jpmml-evaluator/issues/126](url) with your recommendation of using  r2pmml package, however, a new issue occurs: **Some of the ProbabilityDistribution generated by pmml file with jpmml-evaluator are different from the ProbabilityDistribution generated by the original model with R**.

In specific, between model 1(the original model in R and model) and model 2(pmml file translated from 1 using r2pmml library), the ProbabilityDistribution are different (e.g. for "SCL180209114941", model 1 shows 0.4638868 for probability(1) and model 2 shows 0.42524555 for probability(1)). Although the probability difference between them are not large, and prediction result are usually the same, this ProbabilityDistribution difference still shows that something is different between model 1 and 2.  

**Here are the detail prediction result of these two models.**

**These are the results generated by pmml file with jpmml-evaluator(Use Java)**

|          loan_no             | result |    probability(1)     |   probability(0) |
| ---------------------|-------|------------------|----------------|
| SCL180209114941    |     0     |   0.4638868        |     0.5361132    |
| SCL180208113671    |      1     |   0.85728806      |    0.14271194   |
| SCL180225122440   |     0     |   0.049209256    |    0.95079076  |
| SCL180207113563    |     0     |   0.09278628      |    0.9072137     |
| SCL180220119570    |     1      |   0.8018138         |    0.19818622  |
| SCL180222119944    |     0     |   0.450736           |    0.549264     |
| SCL180204109327    |     1     |   0.55664504      |    0.44335496 |
| SCL180207112633    |      0    |   0.23626612       |    0.76373386  |
| SCL180207113540    |      0    |   0.09253664      |    0.9074634    |
| SCL180207112643    |      0    |   0.27287263      |     0.7271274     |
| SCL180215118989    |      0    |   0.1569735         |     0.8430265    |
| SCL180222119931    |      0    |   0.036140308     |    0.9638597    |
| SCL180224121557    |       1    |   0.646922           |    0.353078      |
| SCL180208113824   |      0     |   0.14189377       |    0.85810626   |
| SCL180227123285   |      1     |   0.7287679         |    0.27123213    |
| SCL180206112074    |      0    |   0.26738346       |   0.73261654   |

**These are the results generated by original model with R**

|          loan_no             | result |    probability(1)     |   probability(0) |
| ---------------------|-------|------------------|----------------|
| SCL180209114941    |     0     |   0.42524555      |     0.5747544    |
| SCL180208113671    |      1     |   0.86152035      |    0.1384797      |
| SCL180225122440   |     0     |   0.05436776      |    0.9456322     |
| SCL180207113563    |     0     |   0.10020710       |    0.8997929     |
| SCL180220119570    |     1      |   0.78115392       |    0.2188461      |
| SCL180222119944    |     0     |   0.47037968      |    0.5296203     |
| SCL180204109327    |     1     |   0.58648354      |    0.4135165      |
| SCL180207112633    |      0    |   0.21494596       |    0.7850540     |
| SCL180207113540    |      0    |   0.10329314        |    0.8967069     |
| SCL180207112643    |      0    |   0.24932751        |     0.7506725    |
| SCL180215118989    |      0    |   0.17077683         |     0.8292232   |
| SCL180222119931    |      0    |   0.04397116         |    0.9560288    |
| SCL180224121557    |      1     |   0.61118168          |    0.3888183    |
| SCL180208113824   |      0     |   0.14734288        |    0.8526571     |
| SCL180227123285   |      1     |   0.71817017          |    0.2818298     |
| SCL180206112074    |      0    |   0.27872318         |   0.7212768      |

**Here are my test data (data.csv), pmml file (model.pmml) and original model (original_model.RData) attached.**

[data+two_models.zip](https://github.com/jpmml/jpmml-evaluator/files/2350625/data%2Btwo_models.zip)

**I used the same code as in the last issue [https://github.com/jpmml/jpmml-evaluator/issues/126](url)**

**test class:**
`
import org.dmg.pmml.FieldName;
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Simple {

    public static void main(String[] args) throws IOException {

        String modelPath = "pmml/model.pmml";
        String dataPath = "file/data.csv";
        try {
            ModelInvoker invoker = new ModelInvoker(modelPath);
            List<Map<FieldName, Object>> data = readData(dataPath);
            for (Map<FieldName, Object> elem : data) {
                Map<FieldName, ?> result = invoker.invoke(elem);
                System.out.println(result);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static List<Map<FieldName,Object>> readData(String path) throws IOException{

        BufferedReader br = new BufferedReader(new FileReader(path));
        String[] header = br.readLine().split(",");

        List<Map<FieldName,Object>> data = new ArrayList<>();

        String line = null;
        while((line = br.readLine()) != null){

            Map<FieldName,Object> map = new HashMap<FieldName, Object>();

            String[] dataLine = line.split(",");
            assert (dataLine.length == header.length);
            for(int i=0; i<dataLine.length; i++){
                if(dataLine[i].equalsIgnoreCase("na")) map.put(new FieldName(header[i]), null);
                else map.put(new FieldName(header[i]), dataLine[i]);
            }
            data.add(map);
        }
        return data;
    }
}
`

I have tried to do the following things but the difference between them are still there:
1. remove all missing values(use 0 instead) in both training and testing
2. remove all missing values(use 0.234 instead) in both training and testing
3. remove or add some features(column) in both training and testing

Currently, i have no idea why this happens, so I am hoping for your assistance.

Thank you very much!

ps:
The size of the whole test set is 19432 rows, and 19416 of them are correct (both result and ProbabilityDistribution between pmml file and original model are exactly the same). Only 16 rows listed above are wrong. I only attached this 16 rows above because the whole test set is too large to upload.

**Comment by vruusmann** [comment]
No point in reporting R+XGBoost mispredictions against the JPMML-Evaluator project. The JPMML-Evaluator library is carrying out computations exactly as specified in the PMML document. And I can assure you that the JPMML-Evaluator library knows how to score decision tree models correctly.

The source of the problem is in your R code. Maybe you're handling missing values wrong, or your feature map definition is wrong (eg. mixing `int` and `float` data types etc.).

I personally wouldn't recommend using R as the front-end for XGBoost. You should try Scikit-Learn instead (eg. `xgboost.XGBClassifier`), as it takes much better care of data matrix preparation (eg. missing value handling and feature map generation happen behind the scenes, which eliminates user error component).

**Comment by vruusmann** [comment]
Good way to double-check your R code:
1) Train an XGBoost model, and use it for prediction on training dataset.
2) Store the XGBoost model in a RDS file. Then, restart the R session, and load the XGBoost model from the RDS file.
3) Use the restored model to make predictions on the training dataset again.

Do predictions between steps 1 and 3 agree?

**Comment by vruusmann** [comment]
Also, if you're interested in tracking down the problematic operation in your R code, you should try the following sequence of model training operations (from less complex to more complex data schema):
1. Train the model using non-missing `float` features only. Are R and (J)PMML predictions the same? This should always succeed.
2. Train the model using non-missing `float` and `int` features. If the problem appears now, then your feature map definition is wrong.
3. Train the model using missing and non-missing `float` and `int` features. If the problem appears now, then your missing value handling/encoding is wrong.

**Comment by vruusmann** [comment]
> The most simple model I trained and test only contains int type features.

Try getting started with `float` features - this is the default data type for XGBoost (internally, `int` features are implemented as rounded `float` values, and require a correct feature map definition to work).

TLDR: It's your R code, not XGBoost or (J)PMML code.

**Comment by edenwind** [comment]
I will use only float features and no missing value to try again tomorrow morning.
thank you!

**Comment by vruusmann** [comment]
> pmml's probability distribution can matches R model's

FYI, the "theoretical reproducibility limit" is 1e-6 (three to five ULPs of the 32-bit floating-point datatype), because XGBoost codebase does logit probability calculation dirtily. Also, XGBoost probabilities differ somewhat by version (eg. 0.6 vs 0.64 vs 0.7), because the logit probability calculation is being tweaked constantly there.

 > The int type features are still incorrect, I have double checked the feature map, all features' type in the feature map are int, so the feature map seems to be correct.

Must be your feature map definition.

How did you generate it - manually, or using some helper function/tool? Have you tried the `r2pmml::genFMap()` function? My R2PMML/JPMML-R integration tests are all based on that helper function, and all `float`, `int` and `string` categorical features work flawlessly.

See https://github.com/jpmml/jpmml-r/blob/master/src/test/R/xgboost.R

**Comment by vruusmann** [comment]
> all features' type in the feature map are int

Are you sure that "all features are `int` features" reflects your underlying data schema correctly? Could be that XGBoost is auto-detecting/re-classifying some columns from `int` to `float` (and hence begins to disobey rounding instructions etc.).

**Comment by edenwind** [comment]
Hi

In the case "train the model using non-missing int features only" above, i have already used the r2pmml::genFMap() functin to generate the feature map. the code is below.

This is the R code that i used

```
library(xgboost)
library(data.table)
library(Matrix)
library(dplyr)
library(caret)
library(pmml)
library(CatEncoders)
library(r2pmml)

...
...

xgb.fmap = genFMap(feature_train)
dtrain=xgb.DMatrix(data.matrix(feature_train), label=label_train)
param <- list(objective="binary:logistic",eval_metric = 'auc',eta =0.3,alpha=0.1,max_depth=7,
              colsample_bylevel=1,subsample=1,lambda=50)
model_fit = xgboost(data = dtrain, para=param, nrounds=23)
r2pmml(model_fit, "pmml/model.pmml", fmap = xgb.fmap, response_name = "label",
       response_levels = c("0", "1"),
       compact = TRUE)
```

All features are definitely int type in the "train the model using non-missing int features only" case I mentioned above, because the original data is not int type. I transformed all data to int before training and test. The data was save to csv before training, and I have use the build in search function in MS excel and there was no search result for a dot(.)

**Comment by vruusmann** [comment]
There's still a possibility that your feature map definition and the `dtrain` object are "logically" out of sync. Could you try generating the `xgb.DMatrix` object with the `r2pmml::genDMatrix()` function (and see if that leads to reproducible results)?

Can't remember if the R2PMML package supports the export of Caret-trained XGBoost models or not (the Caret layer would hide the DMatrix and feature map generation from you, and hopefully do it correctly/reproducibly).

**Comment by edenwind** [comment]
Hi
Using the method "r2pmml::genDMatrix()" generates the same result.

This is the new  R code

```
xgb.fmap = genFMap(feature_train)
dtrain = genDMatrix(data.matrix(feature_train), label_train)
param <- list(objective="binary:logistic",eval_metric = 'auc',eta =0.3,alpha=0.1,max_depth=7,
              colsample_bylevel=1,subsample=1,lambda=50)
model_fit = xgboost(data = dtrain, para=param, nrounds=23)
r2pmml(model_fit, "加入人行_Int/shichu_1_clean_noNA.pmml", fmap = xgb.fmap, response_name = "label",
       response_levels = c("0", "1"),
       compact = TRUE)
```

**Comment by vruusmann** [comment]
Does anything change (for the better) if you leave out the `data.matrix()` wrapper function?

The `r2pmml::genDMatrix()` should be able to directly accept a `data.frame` object. If you convert a `data.frame` (every column can have a different data type) to `matrix` (all columns have the same data type), then you're losing quite a bit of useful information.

**Comment by edenwind** [comment]
Hello
Unfortunately, r2pmml::genDMatrix() does not take data frame. It pops up the following error in R.
 
Error in cumsum(col_len) - col_len :
  non-numeric argument to binary operator
 
We have to convert it to matrix to be able to run it. Actually I don’t think this line is the problem, because it has been tested in other cases that genDMatrix is the same as xgb.DMatrix (always same results).

**Comment by vruusmann** [comment]
> Error in cumsum(col_len) - col_len :
> non-numeric argument to binary operator

This is happening here:
https://github.com/jpmml/r2pmml/blob/master/R/xgboost.R#L56

It's a definite proof that your `data.frame` contains heterogeneous columns (esp. categorical integers?), which would give rise to different feature descriptions in a feature map.

Could be that the problem is about distinguishing between categorical and continuous integers in the feature map definition.

> Actually I don’t think this line is the problem, because it has been tested in other cases that genDMatrix is the same as xgb.DMatrix (always same results)

Disagree - this line is exactly the place where I'd be looking for the root cause. 

Take my `r2pmml::genDMatrix()` function, and apply whatever changes necessary to make it work with your `dtrain` object.

When successfully done, please submit a patch against the R2PMML project:
https://github.com/jpmml/r2pmml/issues

**Comment by edenwind** [comment]
Hi there

From last Friday I have tried to find something abnormal data in my training data but nothing was actually wrong. I finally discovered that the reason "**Error in cumsum(col_len) - col_len :
non-numeric argument to binary operator**" occurs when i tried to use data frame as the input parameter for r2pmml::genDMatrix()  is because there was a syntax error." df_X" = and "df_y =" should be added.

I have changed the code as below then no error occurs.

**Before**
```
xgb.fmap = genFMap(feature_train)
dtrain = genDMatrix(feature_train, label_train)
```

**After**
```
xgb.fmap = genFMap(feature_train)
dtrain=genDMatrix(df_X = feature_train, df_y = label_train)
```

**The whole R code is look like this now:**

```
library(xgboost)
library(data.table)
library(Matrix)
library(dplyr)
library(caret)
library(pmml)
library(CatEncoders)
library(r2pmml)
rm(list=ls());gc();gc();cat("\014")
D <- fread("raw_data/Int_2.csv",header =T, data.table = F)
A <- fread("data/data.csv",header =T, data.table = F,
           select=c("loan_no","include_id","t0","min_p","od_days"))
D1 <- D %>% left_join(A,by="loan_no")
feature <- D1[,!(names(D1) %in% c("loan_no","t0","min_p","od_days","include_id"))]
rm(list=ls()[ls()!="D1" & ls()!="feature"]);gc();gc();cat("\014")
label <- as.numeric(!is.na(D1$min_p) & D1$min_p==1)
select_v <- rep(0,length(label))
select_v[D1$t0>17540] <- 1
feature_train <- feature[select_v==1,]
label_train <- label[select_v==1]
filter_v <- D1$min_p[select_v==1]
feature_train <- feature_train[!(!is.na(filter_v) & filter_v>1),]
label_train <- label_train[!(!is.na(filter_v) & filter_v>1)]
feature_test <- feature[select_v<1,]
label_test <- label[select_v<1]
xgb.fmap = genFMap(feature_train)
dtrain=genDMatrix(df_X = feature_train, df_y = label_train)
param <- list(objective="binary:logistic",eval_metric = 'auc',eta =0.3,alpha=0.1,max_depth=7,
              colsample_bylevel=1,subsample=1,lambda=50)
model_fit = xgboost(data = dtrain, para=param, nrounds=23)
r2pmml(model_fit, "no_bug_1.pmml", fmap = xgb.fmap, response_name = "label",
       response_levels = c("0", "1"),
       compact = TRUE)
```

Unfortunately, the result of "** leave out the data.matrix() wrapper function**" is still not correct. The probability distribution between this R model (trained with data frame) and the pmml generated by this R model using r2pmml are still not match for all test data. 

 Using dMatrix before: 16 out of 19432 are not match
 Using data frame now: 19 out of 19432 are not match

Ps: Sorry for bothering you for a long time. It's enough for me to using all float features. If you are busy, just ignore this issue.

Thank you very much!

**Comment by vruusmann** [comment]
> Using dMatrix before: 16 out of 19432 are not match
> Using data frame now: 19 out of 19432 are not match

If you say that the `data.matrix()` function call is doing noting (from the data schema perspective), then adding/removing it shouldn't affect mis-classification numbers?

However, here you have a delta of 3 data records.

> If you are busy, just ignore this issue.

I'm kind of interested of understanding what is going wrong where, so that future versions of the R2PMML package could warn the user.

Have you considered my earlier suggestion that you should try to translate your R code to Scikit-Learn code, and see if the SkLearn2PMML package is giving you a fully reproducible model? Scikit-Learn does not require you to mess with feature map definitions, or column data types explicitly.
